### PR TITLE
Adds Bickley and Scharr gradient filter and improves imgradient docs

### PR DESCRIPTION
**Pull request by zygmuntszpak**

### CODE REVIEW

**Comment by zygmuntszpak** [comment]
Only the nightly builds failed, but presumably that is not a problem. 

The documentation is best read with Juno's new documentation browser which renders LaTex. 
Not sure why, but I am unable to access the documentation for imgradients in IJulia. Perhaps someone else can test it on IJulia?

**Comment by zygmuntszpak** [comment]
@Evizero When you have a moment  would it be possible for you to review this pull-request?

File: src/specialty.jl [hunk 1]
```diff
@@ -17,26 +17,485 @@ end
...
+
+# Determine the mean absolute deviation between the estimated and true
+# orientation. Ignore the values at the border since we expect them to be
+# erroneous.
+error = mean(abs.(direction_true[2:end-1,2:end-1] -
+                 direction_estimated[2:end-1,2:end-1]));
+
+error = round(error,5);
+println("Using \$kernelfunc results in a mean absolute deviation of \$error")
+end
```
Comment (kmsquire) [inline src/specialty.jl]: Consider indenting the code within the `for`-loop.

File: src/specialty.jl [hunk 2]
```diff
@@ -17,26 +17,485 @@ end
...
+Using ImageFiltering.KernelFactors.scharr results in a mean absolute deviation of 0.00126
+Using ImageFiltering.KernelFactors.bickley results in a mean absolute deviation of 0.00038
+```
+# References
+  1. Bernd, J. 2012. *Digital Image Processing* (7th ed.). Springer Publishing Company, Incorporated
+  2. Patra, M. and Karttunen, M. (2006), *Stencils with isotropic discretization error for differential operators*. Numer. Methods Partial Differential Eq., 22:936–953. doi:10.1002/num.20129
+  3. Belyaev, A. (2013). *Implicit Image Differentiation and Filtering with Applications to Image Sharpening*. SIAM Journal on Imaging Sciences, 6(1), 660–679. doi:10.1137/12087092x
+  4. Bickley, W. G. (1948), *Finite difference Formulae for the square lattice*. The Quarterly Journal of Mechanics and Applied Mathematics, 1(1), 35–42. doi:10.1093/qjmam/1.1.35
+  5. Scharr, H., and Weickert, J. (2000). *An Anisotropic Diffusion Algorithm with Optimized Rotation Invariance*. Mustererkennung 2000, 460–467. doi:10.1007/978-3-642-59802-9_58
+---
```
Comment (kmsquire) [inline src/specialty.jl]: The full reference for the Ando kernels is

```
Ando, S. (2000), *Consistent gradient operators*, IEEE Transactions on Pattern Analysis and Machine Intelligence, 22(3), 252-265. doi: 10.1109/34.841757
```

Can you add it here and add a citation above?

File: test/gradient.jl [hunk 3]
```diff
@@ -9,9 +9,10 @@ using Base.Test, Compat  # Compat for ones(x, T)
             (map(v->RGB(v,0,0), y.*ones(x, Float64)), RGB(1,0,0), RGB(0,0,0)))
     for (img, ey, ex) in img_grads
         for kernelfunc in (KernelFactors.ando3, KernelFactors.sobel, KernelFactors.prewitt,
-                           KernelFactors.ando4, KernelFactors.ando5,
+                           KernelFactors.ando4, KernelFactors.ando5,KernelFactors.bickley,
+                           KernelFactors.scharr,
                            Kernel.ando3, Kernel.sobel, Kernel.prewitt,
-                           Kernel.ando4, Kernel.ando5)
+                           Kernel.ando4, Kernel.ando5,Kernel.scharr, Kernel.bickley)
```
Comment (kmsquire) [inline test/gradient.jl]: Nitpick: space before `Kernel.scharr` and `KernelFactors.bickley`.

**Comment by zygmuntszpak** [comment]
Thank you for the positive feedback and the suggestions. I have incorporated all of the recommendations in the latest commit. 

Out of curiosity, how did you become aware of this pull request? Do you get a notification when I make a pull-request, or does only Tim Holy receive it? Are certain people in charge of particular image packages? 

I will continue working on documenting various functions and was wondering if there is anyone in particular I should be requesting a review from? For example, I messaged Tim on #54 but is that considered impolite here?

**Comment by kmsquire** [comment]
On Wed, Jan 17, 2018 at 2:56 AM Dr. Zygmunt L. Szpak <
notifications@github.com> wrote:

> Thank you for the positive feedback and the suggestions. I have
> incorporated all of the recommendations in the latest commit.
>
You're very welcome!

> Out of curiosity, how did you become aware of this pull request? Do you
> get a notification when I make a pull-request, or does only Tim Holy
> receive it? Are certain people in charge of particular image packages?
>
I’m an infrequent contributor to Images.jl, and I receive email
notifications for all issues and pull requests.
This is probably true of anyone with commit access, although those people
may choose not to receive messages.
Tim has been the driving force behind the various Image packages, although
various others contribute more or less, depending on the package.
I don’t have much time for Julia these days, but I generally follow what’s
happening and contribute when I can.
This particular pull request piqued my interest because I contributed the
ando methods.

> I will continue working on documenting various functions and was wondering
> if there is anyone in particular I should be requesting a review from? For
> example, I messaged Tim on #54
> <https://github.com/JuliaImages/ImageFiltering.jl/pull/54> but is that
> considered impolite here?
>
Not impolite at all!  Tim, in particular, is very prolific in Julia in
general, and is generally quite good at responding to people, but like most
of us, he's also a very busy person, so it might take him some time to
respond.
If you find a pull request languishing for a few days without an answer,
you should feel free to bump it.  Even if Tim is not available, someone
else will probably make the time to review it.

File: src/specialty.jl [hunk 4]
```diff
@@ -17,26 +17,486 @@ end
...
+    \\begin{bmatrix}
+    -1 & 0 & 1
+    \\end{bmatrix}.
+\\end{aligned}
+```
+See also: [`sobel`](@ref), [`KernelFactors.sobel`](@ref) and
+[`Kernel.sobel`](@ref)
+
+### `KernelFactors.ando3`
+The *ando3* option specifies the kernels
```
Comment (kmsquire) [inline src/specialty.jl]: For consistency with *scharr* and *bickley*, 
```
The *ando3* option [6] specifies the kernels
```
(although you might want to renumber the references)

Comment (zygmuntszpak) [reply src/specialty.jl]: Done

**Comment by kmsquire** [comment]
@zygmuntszpak, when I render the document, I'm getting a number of unknown references related to changes here:

```julia
$ julia make.jl
Documenter: setting up build directory.
Documenter: expanding markdown templates.
Documenter: building cross-references.
 !! No doc found for reference '[`prewitt`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`sobel`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`ando3`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`ando4`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`ando5`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`scharr`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`KernelFactors.scharr`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`Kernel.scharr`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`bickley`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`KernelFactors.bickley`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`Kernel.bickley`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`Kernel.scharr`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`Kernel.bickley`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`Kernel.scharr`](@ref)'. [src/function_reference.md]
 !! No doc found for reference '[`Kernel.bickley`](@ref)'. [src/function_reference.md]
 !! Unable to get the binding for '[`Kernel.ando3`](@ref)'. [src/function_reference.md]

ERROR: UndefVarError(:Kernel)

expression 'Kernel.ando3' in module 'ImageFiltering.KernelFactors'

 !! Unable to get the binding for '[`Kernel.ando4`](@ref)'. [src/function_reference.md]

ERROR: UndefVarError(:Kernel)

expression 'Kernel.ando4' in module 'ImageFiltering.KernelFactors'

 !! Unable to get the binding for '[`Kernel.ando5`](@ref)'. [src/function_reference.md]

ERROR: UndefVarError(:Kernel)

expression 'Kernel.ando5' in module 'ImageFiltering.KernelFactors'

 !! No doc found for reference '[`Kernel`](@ref)'. [src/index.md]
 !! No doc found for reference '[`KernelFactors`](@ref)'. [src/index.md]
```

If you have time to fix these, that would be great, although to keep things moving, those changes can happen in a future pull request (by you or someone else) if you prefer.  

Either way, until they're fixed, a few of the documentation links (related to the errors above) won't work, but even in its current state, I really think that both the documentation and code additions are great, and would love to get them in.  Let me know

**Comment by zygmuntszpak** [comment]
Thanks for looking into it more deeply. I didn't actually know how to compile the docs. I was using IJulia and the Atom documentation browser to check things. In IJulia all of the reference links don't work, but that seems to be the case for everything not just the Images package. 

I've managed to compile the documentation now, so I will work on fixing the broken links over the weekend.

**Comment by zygmuntszpak** [comment]
I've managed to resolve the unknown references that occur when attempting to build the documentation. I've also made the requisite change to avoid the deprecation warning that was being triggered. 

Here is the log of my build:

> 

julia> include("make.jl")
Documenter: setting up build directory.
Documenter: expanding markdown templates.
Documenter: building cross-references.
Documenter: running document checks.
 > checking for missing docstrings.
 !! 9 docstrings potentially missing:

    ImageFiltering.MapWindow.extrema_filter :: Union{Tuple{AbstractArray{T,N},Tuple{Vararg{Integer,N}}}, Tuple{N}, Tuple{T}} where N where T
    ImageFiltering.shrink :: Tuple{Tuple{Vararg{AbstractUnitRange,N}} where N,Any}
    ImageFiltering.expand :: Tuple{Tuple{Vararg{AbstractUnitRange,N}} where N,Any}
    ImageFiltering.map_copytail :: Tuple{Any,Tuple,Tuple}
    ImageFiltering.KernelFactors.iterdims :: Union{Tuple{Npre}, Tuple{N}, Tuple{Tuple{Vararg{AbstractUnitRange,N}},ImageFiltering.KernelFactors.ReshapedOneD{T,N,Npre,V} where V}, Tuple{T}} where Npre where N where T
    ImageFiltering.padindex :: Tuple{ImageFiltering.Pad,Integer,AbstractUnitRange,Integer}
    ImageFiltering.p :: Tuple{Any}
    ImageFiltering.Algorithm.FIRTiled
    ImageFiltering.MapWindow.mapwindow! :: Union{NTuple{4,Any}, NTuple{5,Any}, NTuple{6,Any}}

 > running doctests.
 > checking footnote links.
Documenter: populating indices.
Documenter: rendering document.
Documenter: skipping docs deployment.
  You can set DOCUMENTER_DEBUG to "true" in Travis to see more information.


It would be worth checking that you get the same results on your machine.

**Comment by kmsquire** [comment]
Looks good!

> @kmsquire I'm not sure if you get notifications everytime I post a comment on this thread, so I'll mention you directly just in case.

I do get all notifications, but you should feel free to ping me anyway--I often can't review things right away, and lead a busy enough life that I'll forget.

**Comment by timholy** [comment]
Thanks @kmsquire for handling this!

**Pull request merged**